### PR TITLE
Fix Auto Crafter Empty Slot Check

### PR DESCRIPTION
**Pull request by LilBC**
## Description

There is an empty slot checking bug in the auto crafters. They only check if there is a slot in the chest that is completely empty. But if there is a stack of item in the chest that is not full but same item as the result of the recipe, the auto crafter will stop crafting because it thinks there are no empty slots.

## Proposed changes

I added a new method in the Abstract Auto Crafter file that checks if there is a stack of item that is the same type if there are no slots that are completely empty.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java [hunk 1]
```diff
@@ -391,6 +392,29 @@ protected boolean matchesAny(Inventory inv, Map<Integer, Integer> itemQuantities
...
+
+        if(inv.firstEmpty() < 0){
+
+            for (ItemStack matchingItem : inv.getContents()) {
+                // Checks if any item stack that is not full that is the same type as the result.
+                if (
+                        ((!(resultItem instanceof SlimefunItemStack) && resultItem.isSimilar(matchingItem))
+                        || ((resultItem instanceof SlimefunItemStack) && SlimefunUtils.isItemSimilar(matchingItem, resultItem, true)))
+                        && matchingItem.getAmount() < matchingItem.getMaxStackSize()
+                ){
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java:405]: formatting :notLikeThis:

**Comment by martinbrom** [comment]
Duplicate of https://github.com/Slimefun/Slimefun4/issues/2901

**Comment by martinbrom** [comment]
@LilBC couldnt you use InvUtils.fits() instead?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java (hunk 1 above)
Comment (LilBC) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java:405]: Ops, sorry. I read the code style guidelines, I just didn't know what does line-wrapping means.

**Comment by LilBC** [comment]
> Duplicate of #2901
> 
> I wanted to implement this as soon as the new autocrafters came out but 🍪 was against it. Felt like it would simplify using the autocrafters...
> 
> EDIT: thebusybot marked as duplicate even though I am not a member??

Ooops, my bad. I should've searched for this issue before I start coding. I guess I'll wait for 🍪 to see this and see what do they think... Probably gonna get closed. Sad... wasted too much time coding. :(

**Comment by LilBC** [comment]
> @LilBC couldnt you use InvUtils.fits() instead?

Cool, I didn't know this method existed. I'll try to change to using it.

**Comment by martinbrom** [comment]
> Edit: Hmm... I don't think it would work... I think it only works on Dirty Chest Menus ...

Not true, the contract of the method is `boolean fits(@NonNull Inventory inv, @NonNull ItemStack item, int... slots)`

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java [hunk 2]
```diff
@@ -414,8 +434,8 @@ public boolean craft(@Nonnull Inventory inv, @Nonnull AbstractRecipe recipe) {
             return false;
         }
 
-        // Check if we have an empty slot
-        if (inv.firstEmpty() != -1) {
+        // Check if the inventory have any available space
+        if (inventoryAvailable(inv, recipe.getResult())) {
```
Comment (martinbrom) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java:438]:
```suggestion
        if (InvUtils.fits(inv, recipe.getResult())) {
```

**Review by martinbrom — COMMENTED** [review]
This would be the only change you'd need to do in case :cookie: changed his mind.
Also would need to add the import manually, the suggestions don't do that on their own.

Your code wouldn't work if the recipe produced more than 1 item. It's usually better to use existing code instead of writing everything yourself.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java [hunk 3]
```diff
@@ -391,6 +392,25 @@ protected boolean matchesAny(Inventory inv, Map<Integer, Integer> itemQuantities
...
+        if(inv.firstEmpty() < 0){
+
+            for (ItemStack matchingItem : inv.getContents()) {
+                // Checks if any item stack that is not full that is the same type as the result.
+                if (((!(resultItem instanceof SlimefunItemStack) && resultItem.isSimilar(matchingItem)) || ((resultItem instanceof SlimefunItemStack) && SlimefunUtils.isItemSimilar(matchingItem, resultItem, true))) && matchingItem.getAmount() < matchingItem.getMaxStackSize()){
+                    return true;
+                }
+            }
+            return false;
+        }
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java:406]:
```suggestion
        if (inv.firstEmpty() < 0) {

            for (ItemStack matchingItem : inv.getContents()) {
                // Checks if any item stack that is not full that is the same type as the result.
                if (((!(resultItem instanceof SlimefunItemStack) && resultItem.isSimilar(matchingItem)) 
                	|| ((resultItem instanceof SlimefunItemStack) && SlimefunUtils.isItemSimilar(matchingItem, resultItem, true))) 
                	&& matchingItem.getAmount() < matchingItem.getMaxStackSize()
                ){
                    return true;
                }
            }
            return false;
        }
```

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

**Comment by TheBusyBiscuit** [comment]
Rejecting this for the same reason as last time.
It's intended to work this way. Item comparisons are very expensive and this change would turn Auto-Crafters into absolute lag machines.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java (hunk 2 above)
Comment (LilBC) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/autocrafters/AbstractAutoCrafter.java:438]: Ohhh ~ Okay! Thank you. Sorry I'm just kinda noob I didn't know how to use this method. Very noice.

**Comment by LilBC** [comment]
> Rejecting this for the same reason as last time.
> It's intended to work this way. Item comparisons are very expensive and this change would turn Auto-Crafters into absolute lag machines.

Can I do it like after the player sets and saves the recipe, the autocrafter searches for and saves the first same item slot to only check that slot every tick. So it doesn't searchs for it and compare all of the items every tick. I think this feature is pretty important for automation since it makes automation so much easier.

**Pull request closed without merging**